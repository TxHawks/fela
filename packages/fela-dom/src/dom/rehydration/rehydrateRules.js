/* @flow */
import camelCaseProperty from 'css-in-js-utils/lib/camelCaseProperty'
import { generateCSSSelector, RULE_TYPE } from 'fela-utils'

import type DOMRenderer from '../../../../../flowtypes/DOMRenderer'

const DECL_REGEX = /.([^:{]+)(:[^{]+)?{([^}]+)}/g
const PROPERTY_VALUE_REGEX = /:(.+)/

export default function rehydrateRules(
  renderer: DOMRenderer,
  cache: Object,
  css: string,
  media: string = '',
  support: string = ''
) {
  let decl

  // This excellent parsing implementation was originally taken from Styletron and modified to fit Fela
  // https://github.com/rtsao/styletron/blob/master/packages/styletron-client/src/index.js#L47
  /* eslint-disable no-unused-vars,no-cond-assign */
  while ((decl = DECL_REGEX.exec(css))) {
    // $FlowFixMe
    const [ruleSet, className, pseudo = '', declaration] = decl
    /* eslint-enable */
    const [property, value] = declaration.split(PROPERTY_VALUE_REGEX)

    renderer.getNextRuleIdentifier()

    const declarationReference =
      support + media + pseudo + camelCaseProperty(property) + value
    cache[declarationReference] = {
      type: RULE_TYPE,
      className,
      selector: generateCSSSelector(className, pseudo),
      declaration,
      media,
      support
    }
  }
}
