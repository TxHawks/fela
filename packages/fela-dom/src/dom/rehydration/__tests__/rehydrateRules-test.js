import rehydrateRules from '../rehydrateRules'

describe('Rehydrating rules', () => {
  it('should rehydrate the renderer cache', () => {
    expect(
      rehydrateRules(
        '.a{color:red}.b{color:blue}.c:hover:active{border-color:2px solid rgb(255, 255, 0)}'
      )
    ).toMatchSnapshot()
  })
})
