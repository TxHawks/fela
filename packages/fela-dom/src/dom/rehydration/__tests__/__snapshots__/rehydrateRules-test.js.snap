// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rehydrating rules should rehydrate the renderer cache 1`] = `
Object {
  ":hover:activeborderColor2px solid rgb(255, 255, 0)": Object {
    "className": "c",
    "declaration": "border-color:2px solid rgb(255, 255, 0)",
    "media": "",
    "selector": ".c:hover:active",
    "support": "",
    "type": "RULE",
  },
  "colorblue": Object {
    "className": "b",
    "declaration": "color:blue",
    "media": "",
    "selector": ".b",
    "support": "",
    "type": "RULE",
  },
  "colorred": Object {
    "className": "a",
    "declaration": "color:red",
    "media": "",
    "selector": ".a",
    "support": "",
    "type": "RULE",
  },
}
`;
